src/Http/Http/src/Features/QueryFeature.cs

```diff
@@ -100,5 +102,237 @@ public IQueryCollection Query
+                    {
+                        if (!char.IsWhiteSpace(querySegment[i]))
+                        {
+                            break;
+                        }
+                    }
+
+                    var name = querySegment[i..equalIndex];
+                    var value = querySegment.Slice(equalIndex + 1);
+
+                    SpanHelper.ReplacePlusWithSpaceInPlace(name);
+                    SpanHelper.ReplacePlusWithSpaceInPlace(value);
+
+                    accumulator.Append(
+                        Uri.UnescapeDataString(name.ToString()),
+                        Uri.UnescapeDataString(value.ToString()));
+                }
+                else
+                {
+                    accumulator.Append(querySegment);
+                }
+
+                if (delimiterIndex < 0)
+                {
+                    break;
+                }
+
+                query = query.Slice(delimiterIndex + 1);
+            }
+
+            if (arryToReturnToPool is not null)
+            {
+                ArrayPool<char>.Shared.Return(arryToReturnToPool);
+            }
+
+            if (!accumulator.HasValues)
+            {
+                return null;
+            }
+
+            return accumulator.GetResults();
+        }
+
+        internal struct KvpAccumulator
+        {
+            /// <summary>
+            /// This API supports infrastructure and is not intended to be used
+            /// directly from your code. This API may change or be removed in future releases.
+            /// </summary>
+            private AdaptiveCapacityDictionary<string, StringValues> _accumulator;
+            private AdaptiveCapacityDictionary<string, List<string>> _expandingAccumulator;
+
+            public void Append(ReadOnlySpan<char> key, ReadOnlySpan<char> value = default)
+                => Append(key.ToString(), value.IsEmpty ? string.Empty : value.ToString());
+
+            /// <summary>
+            /// This API supports infrastructure and is not intended to be used
+            /// directly from your code. This API may change or be removed in future releases.
+            /// </summary>
+            public void Append(string key, string value)
+            {
+                if (key.Length == 0)
+                {
+                    return;
+                }
+
+                if (_accumulator is null)
+                {
+                    _accumulator = new AdaptiveCapacityDictionary<string, StringValues>(StringComparer.OrdinalIgnoreCase);
+                }
+
+                if (_accumulator.TryGetValue(key, out var values))
+                {
+                    if (values.Count == 0)
+                    {
+                        // Marker entry for this key to indicate entry already in expanding list dictionary
+                        _expandingAccumulator[key].Add(value);
+                    }
+                    else if (values.Count == 1)
```

Comment:
I can understand why you'd special case single values as very common, but why is the second value special compared to 3+?

Reply:
I can't see what you mean, especially 

> why is the second value special compared to 3+

Please give me a pointer.
(Note: this part of the PR is from @jkotalik)

Reply:
The first time you find a key you hit the else block down on line 250 and you make a new StringValues. The second time you see that key you come to this if block (227) and Concat onto the first StringValues. The third time you end up in the else block on line 231 and switch over to the _expandingAccumulator. Any subsequent hits for the same key end up in the if block on line 222 and append to the accumulator. (Maybe this code should be re-ordered to follow that logical flow 😁).

The question I was asking was why is step 2 with the Concat special? You could remove it and go directly to using the accumulator. Having a single query value seems like the primary case to optimize for, once you have multiple you might as well use the accumulator.

Reply:
Got you, thanks.
I don't why it was chosen to be that way, but I agree with you, that this can be simplified.

Reply:
Changed the order, and split Append into two parts. One (fast-path) for single query value, and a cold-path for 2+ values.
bbcda61981e578faecdfce1f167eb2990a9cc3b2